contracts/governance/Governor.sol

```diff
@@ -0,0 +1,102 @@
+pragma solidity ^0.5.11;
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+/**
+ * @title Governor
+ * @dev The Governor holds the rights to stage and execute contract calls i.e. changing Livepeer protocol parameters. 
+ */
+contract Governor {
+
+    address public owner;
+
+    uint256 public nonce;
+
+    /// @dev mapping of updateHash (keccak256(update) => executeBlock (block.number + delay)
+    mapping(bytes32 => uint256) updates; 
+
+    using SafeMath for uint256;
+
+    event OwnershipTransferred(address indexed previousOwner, address indexed newOwner);
+
+    event UpdateStaged(Update update, uint256 delay);
+
+    event UpdateExecuted(Update update);
+
+    struct Update {
+        address[] target;
+        uint256[] value;
+        bytes[] data;
+        uint256 nonce;
+    }
+
+/**
+   * @notice Throws if called by any account other than the owner.
+   */
+    modifier onlyOwner() {
+        require(msg.sender == owner, "unauthorized");
+        _;
+    }
+
+
+    /// @notice Throws if called by any account other than this contract.
+    /// @dev Forces the `stage/execute` path to be used to call functions with this modifier instead of directly.
+    modifier onlyThis() {
+        require(msg.sender == address(this), "unauthorized");
+        _;
+    }
+
+
+    /// @dev The Ownable constructor sets the original `owner` of the contract to the sender account.
+    constructor() public {
+        owner = msg.sender;
+    }
+
+
+    /// @notice Allows the current owner to transfer control of the contract to a newOwner.
+    /// @dev Can only be called through stage/execute, will revert if the caller is not this contract's address.
+    /// @param newOwner The address to transfer ownership to.
+
+    function transferOwnership(address newOwner) public onlyThis {
+        require(newOwner != address(0));
+        owner = newOwner;
+        emit OwnershipTransferred(owner, newOwner);
+    }
+    
+    /// @notice Stage a batch of updates to be executed
+    /// @dev Anyone can execute this but only transactions signed by the owner are valid
+    /// @param _update Update to be staged
+    /// @param _delay (uint256) Delay (in number of blocks) for the update
+    function stage(Update memory _update, uint256 _delay) public onlyOwner {
+        require(nonce < _update.nonce, "nonce too low");
```

Comment:
Right, this is not as useful without signature based auth. 

I guess it's useful as an index for the edge case where there's multiple updates in a single block, otherwise blocknumbers would suffice to have historical ordering.

Reply:
The `UpdateStaged` and `UpdateExecuted` events within a single block should be emitted in the order that updates are staged or executed. I believe this ordering is achieved by attaching an index to each log within a block (would be good to double check this).

Furthermore, I think having a nonce included in the update makes sense, but I don't think the contract needs to track a global nonce and do any checks on the nonce included in an update.

Reply:
Ordering of transactions within a single block are determined by the miner based on transaction nonce and price (and can thus vary between implementations) thus I don't think there's guarantees that log index (or nonce tracked client-side by events) ordering is preserved. 

The nonce in this contract does provide that guarantee, so I think the question is whether the tradeoff of a `SLOAD` and `STORE` against having hard ordering is worth it. 

The way the delay mapping works in this implementation should provide ample replay and re-entry prevention (it would be good to have this as a highlight for the audit) so it's more a nice to have rather than an explicit security requirement.

Reply:
Hm I actually think regular transaction nonces already handle the scenario that you're describing.

2 distinct `stage()`/`update()` calls can be included in the same block in the following scenarios:

1. If the contract owner batched the calls together into a single atomic transaction
2. If the contract owner made two separate calls in two separate transactions and those transactions happen to be included in the same block

In scenario 1, the ordering of the calls is defined by the signer of the transaction.

In scenario 2, regardless of whether the contract owner itself is a contract or an EOA, each of the transactions would have a nonce that guarantees execution order - the transaction that was created/signed first should have a lower nonce and should be executed before the second transaction.

Reply:
In scenario 1 while true, other stakeholders depend on the benevolence of the `owner` to define the right nonces/indexes. There are no barriers of submitting multiple transactions with the same nonce or older nonces if it is not tracked and checked. 

Full transparency does come at a cost, in this case that of a storage read and write.

Scenario 2 does only uphold for EOA's because contract accounts will not increment nonces unless for `CREATE` (and I think `CREATE2` opcodes) 

From the [Ethereum yellow paper](http://gavwood.com/paper.pdf), page 3, section 4.1 'World State' :  

> nonce: A scalar value equal to the number of transactions sent from this address or, in the case
of accounts with associated code, the number of
contract-creations made by this account. For account of address a in state σ, this would be formally denoted σ[a]n.

Reply:
> There are no barriers of submitting multiple transactions with the same nonce or older nonces if it is not tracked and checked.

Right, but the contract owner can always submit the same update with a nonce that passes the contract nonce check. So, I'm not sure that I see what the contract nonce check protects against in this scenario. Since the contract owner has the authority to call `stage()`, it can stage whatever update that it wants.

> Scenario 2 does only uphold for EOA's because contract accounts will not increment nonces

The scenario I am referring to here involves a contract calling the Governor in two separate transactions (the contract calls against the Governor would be "internal transactions", but those shouldn't be relevant here since we just care about the other transactions that triggered the contract calls in the first place). In order for there to be two separate transactions, an EOA would have to submit two separate transactions to the contract. Each of these transactions will need to have a unique nonce.

Reply:
> Right, but the contract owner can always submit the same update with a nonce that passes the contract nonce check.

That's true but the owner can't submit an update with the same nonce in the same block. 

The owner could theoretically submit a batch of `stage()` calls within the same transaction using a smart contract wallet. 

I'm okay with removing the nonce notion since it's unlikely this scenario will ever take place though, but let's be mindful of it.

Reply:
removed notion of a nonce in 9050b40